File: examples/item-list/src/main/java/com/gurunars/item_list/example/AnimalItem.java
package com.gurunars.item_list.example;

import com.gurunars.item_list.Item;

class AnimalItem extends Item<AnimalItem.Type> {

    private int version;

    @Override
    public boolean payloadsEqual(Item other) {
        return other instanceof AnimalItem && version == ((AnimalItem) other).version;
    }

    enum Type {
        MONKEY, TIGER, WOLF, LION
    }

    void update() {
        this.version++;
    }

    AnimalItem(long id, int version, Type type) {
        super(id, type);
        this.version = version;
    }

    @Override
    public String toString() {
        return "#" + getId() + "{" + getType() + " @ " + version + "}";
    }

}

File: examples/selectable-item-list/src/main/java/com/gurunars/item_list/selectable_example/AnimalItem.kt
package com.gurunars.item_list.selectable_example

import com.gurunars.item_list.Item

internal class AnimalItem(private var id: Long, private var version: Int, private var type: AnimalItem.Type) : Item {

    override fun getType(): Enum<*> {
        return type
    }

    override fun getId(): Long {
        return id
    }

    override fun payloadsEqual(other: Item): Boolean {
        return other is AnimalItem && version == other.version
    }

    internal enum class Type {
        MONKEY, TIGER, WOLF, LION
    }

    fun update() {
        this.version++
    }

    override fun toString(): String {
        return "#$id{$type @ $version}"
    }

}